### PR TITLE
Adding strained crystal method to crystal

### PR DESCRIPTION
**Pull request by cophus**
crystal_strained = crystal.get_strained_crystal(
        exx = 0.0,
        eyy = 0.0,
        ezz = 0.0,
        exy = 0.0,
        exz = 0.0,
        eyz = 0.0,
)

is the syntax

### CODE REVIEW

File: py4DSTEM/process/diffraction/crystal_bloch.py [hunk 1]
```diff
@@ -74,6 +74,8 @@ def calculate_dynamical_structure_factors(
                                         substantial speedup at the cost of some reduced accuracy
 
         See WK_scattering_factors.py for details on the Weickenmeier-Kohl form factors.
+
+
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal_bloch.py:78]: remove whitespace

File: py4DSTEM/process/diffraction/crystal.py [hunk 2]
```diff
@@ -386,13 +459,28 @@ def calculate_structure_factors(
         k_max: float = 2.0,
         tol_structure_factor: float = 1e-4,
         return_intensities: bool = False,
-    ):
+        ):
+
+
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:464]: whitespace

File: py4DSTEM/process/diffraction/crystal.py [hunk 3]
```diff
@@ -386,13 +459,28 @@ def calculate_structure_factors(
...
+            tolerance for removing low-valued structure factors
+        return_intensities: bool
+            return the intensities and positions of all structure factor peaks.
+        
+        Returns
+        --------
+        (q_SF, I_SF)
+            Tuple of the q vectors and intensities of each structure factor.
+        
+
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:483]: whitespace

File: py4DSTEM/process/diffraction/crystal.py [hunk 4]
```diff
@@ -139,6 +149,69 @@ def calculate_lattice(self):
...
+        exy = 0.0,
+        exz = 0.0,
+        eyz = 0.0,
+        deformation_matrix = None,
+        return_deformation_matrix = False,
+        ):
+        """
+        This method returns new Crystal class with strain applied. The directions of (x,y,z)
+        are with respect to the default Crystal orientation, which can be checked with
+        print(Crystal.lat_real)
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:167]: Should we add which notation we expect users to pass or the meaning of the floats i.e. 0.01 is a 1% strain

File: py4DSTEM/process/diffraction/crystal.py [hunk 5]
```diff
@@ -170,9 +170,9 @@ def get_strained_crystal(
         # deformation matrix
         if deformation_matrix is None:
             deformation_matrix = np.array([
-                [1+exx,     exy,    exz     ],
-                [exy,       1+eyy,  eyz     ],
-                [exz,       eyz,    1+ezz   ],
+                [1.0+exx,   1.0*exy,    1.0*exz],
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:173]: add comment as to which convention this follows?

File: py4DSTEM/process/diffraction/crystal.py [hunk 6]
```diff
@@ -108,7 +109,16 @@ def __init__(
         
         # Calculate lattice parameters
         self.calculate_lattice()
-        
+    
+    # TODO check for and copy all important attributes
+    def copy(self):
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:114]: Do we expect users to do this, if not we should name it `_copy()` as it's not really a copy but init a new crystal,

**Review by alex-rakowski — APPROVED** [review]
Looks correct to me. Few formatting issues, and I think the copy method should be changed to `_copy()` or given a different name.

**Comment by sezelt** [comment]
I removed the copy method since it is no longer used, and under some conditions it would not produce an exact copy of the crystal (i.e. if the Crystal was generated with the 3x3 lattice specification, the copy would have the correct cell but could have rotated axes)

File: py4DSTEM/process/diffraction/crystal.py [hunk 7]
```diff
@@ -139,6 +156,49 @@ def calculate_lattice(self):
...
+        """
+        This method returns new Crystal class with strain applied. The directions of (x,y,z)
+        are with respect to the default Crystal orientation, which can be checked with
+        print(Crystal.lat_real) applied to the original Crystal.
+
+        Strains are given in fractional values, so exx = 0.01 is 1% strain along the x direction.
+        """
+
+        # deformation matrix
+        if deformation_matrix is None:
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:180]:
```suggestion
        # pymatgen flag
        if 'pymatgen' in sys.modules:
            self.pymatgen_available = True            
        else:
            self.pymatgen_available = False
```

File: py4DSTEM/process/diffraction/crystal.py [hunk 8]
```diff
@@ -139,6 +156,49 @@ def calculate_lattice(self):
             self.pymatgen_available = True            
         else:
             self.pymatgen_available = False
```
Comment (alex-rakowski) [inline py4DSTEM/process/diffraction/crystal.py:158]: Move this to outside calc_lattice, to just in the init? its not a lattice calc

**Pull request merged**